primeorder/src/point_arithmetic.rs

```diff
@@ -0,0 +1,327 @@
+//! Point arithmetic implementation optimised for different curve equations
+//!
+//! Support for formulas specialized to the short Weierstrass equation's
+//! 𝒂-coefficient.
+//!
+//! This module is largely a workaround for things which should be possible
+//! to implement more elegantly with future Rust features like
+//! `generic_const_exprs` and `impl const Trait`.
+//!
+//! In absence of such features, we define ZSTs that implement point arithmetic
+//! for different curve equations that depend on properties of the 𝒂-coefficient
+//! which could potentially be written as const expressions on `PrimeCurveParams::EQUATION_A`
+//! in the future (including ones which could be used as trait bounds).
+
+use elliptic_curve::{subtle::ConditionallySelectable, Field};
+
+use crate::{AffinePoint, PrimeCurveParams, ProjectivePoint};
+
+mod sealed {
+    /// Seals `PointArithmetic` trait
+    pub trait Sealed {}
+}
+
+/// Elliptic point artihmetic implementation
+///
+/// Provides implementation of point arithmetic (point addition, point doubling) which
+/// might be optimized for the curve.
+pub trait PointArithmetic<C: PrimeCurveParams>: sealed::Sealed {
+    /// Returns `point + point`
+    fn double(point: &ProjectivePoint<C>) -> ProjectivePoint<C>;
+    /// Returns `lhs + rhs`
+    fn add(lhs: &ProjectivePoint<C>, rhs: &ProjectivePoint<C>) -> ProjectivePoint<C>;
+    /// Returns `lhs + rhs`
+    fn add_mixed(lhs: &ProjectivePoint<C>, rhs: &AffinePoint<C>) -> ProjectivePoint<C>;
+}
```

Comment:
I'm not sure I see the purpose of having two traits, one sealed (with a nondescript name) and one not, when `PointArithmetic` could just be sealed.

If it were sealed, it couldn't be used in bounds, but I'm not sure where that would be necessary.

Reply:
There's indeed no benefit in exposing `PointArithmetic` trait as `ProjectivePoint` is the only one who's supposed to use it. I'll hide entire trait then.

Reply:
Done!